### PR TITLE
feat: enhance Azure deployment code to support Azure Stack.

### PR DESCRIPTION
**Pull request by rjaini**
Currently Azure deployment code only support execution of the test on Azure Cloud.
The changes is to enhance the code to support execution on Azure Stack as well.

/cc @PatrickLang
/cc @ritazh

### CODE REVIEW

**Comment by rjaini** [comment]
cc @feiskyer @jchauncey

File: kubetest/azure.go [hunk 1]
```diff
@@ -101,12 +111,93 @@ type Cluster struct {
...
+// IsAzureStackCloud return true if the cloud is AzureStack
+func (c *Cluster) isAzureStackCloud() bool {
+	return c.azureCustomCloudURL != "" && strings.EqualFold(c.azureEnvironment, AzureStackCloud)
+}
+
+// SetCustomCloudProfileEnvironment retrieves the endpoints from Azure Stack metadata endpoint and sets the values for azure.Environment
+func (c *Cluster) SetCustomCloudProfileEnvironment() error {
+	var environmentJSON string
+	if c.isAzureStackCloud() {
+
```
Comment (feiskyer) [inline kubetest/azure.go]: nit: remove empty line

File: kubetest/azure.go [hunk 2]
```diff
@@ -65,9 +66,18 @@ var (
...
+	acsCustomCloudURL      = flag.String("acsengine-custom-cloud-url", "", "management portal URL to use in custom Azure cloud (i.e Azure Stack etc)")
 	testCcm                = flag.Bool("test-ccm", false, "Set to True if you want kubetest to run e2e tests for ccm")
 )
 
+const (
+	// AzureStackCloud is a const string reference identifier for Azure Stack cloud
+	AzureStackCloud = "AzureStackCloud"
+	// ADFS Identity System for Azure Stack cloud
+	ADFSIdentitySystem = "adfs"
+)
```
Comment (feiskyer) [inline kubetest/azure.go]: nit: add a new empty line after `)`

File: kubetest/azure.go [hunk 3]
```diff
@@ -101,12 +111,93 @@ type Cluster struct {
...
+func (c *Cluster) SetCustomCloudProfileEnvironment() error {
+	var environmentJSON string
+	if c.isAzureStackCloud() {
+
+		env := azure.Environment{}
+		env.Name = c.azureEnvironment
+		azsFQDNSuffix := strings.Replace(c.azureCustomCloudURL, fmt.Sprintf("https://portal.%s.", c.location), "", -1)
+		azsFQDNSuffix = strings.TrimSuffix(azsFQDNSuffix, "/")
+		env.ResourceManagerEndpoint = fmt.Sprintf("https://management.%s.%s/", c.location, azsFQDNSuffix)
+		metadataURL := fmt.Sprintf("%s/metadata/endpoints?api-version=1.0", strings.TrimSuffix(env.ResourceManagerEndpoint, "/"))
```
Comment (feiskyer) [inline kubetest/azure.go]: would this work if the kubetest is running on non-Azure environments?

File: kubetest/azure.go (hunk 1 above)
Comment (rjaini) [reply kubetest/azure.go]: Formatted the code.

File: kubetest/azure.go (hunk 2 above)
Comment (rjaini) [reply kubetest/azure.go]: Formatted the code.

File: kubetest/azure.go (hunk 3 above)
Comment (rjaini) [reply kubetest/azure.go]: Yes. I tested changes and after above changes Kubetest will be able to launched on Azure Stack env.

**Comment by PatrickLang** [comment]
/lgtm
@bclau  - can you do a review too?

**Comment by rjaini** [comment]
@feiskyer : Let me know if you have any more comments.

**Comment by krzyzacy** [comment]
please fix lint, otherwise lgtm

File: kubetest/azure.go [hunk 4]
```diff
@@ -65,9 +66,19 @@ var (
...
+	acsAzureEnv            = flag.String("acsengine-azure-env", "AzurePublicCloud", "The target Azure cloud")
+	acsIdentitySystem      = flag.String("acsengine-identity-system", "azure_ad", "identity system (default:`azure_ad`, `adfs`)")
+	acsCustomCloudURL      = flag.String("acsengine-custom-cloud-url", "", "management portal URL to use in custom Azure cloud (i.e Azure Stack etc)")
 	testCcm                = flag.Bool("test-ccm", false, "Set to True if you want kubetest to run e2e tests for ccm")
 )
 
+const (
+	// AzureStackCloud is a const string reference identifier for Azure Stack cloud
+	AzureStackCloud = "AzureStackCloud"
+	// ADFS Identity System for Azure Stack cloud
```
[the remark this answers is not in the document]
Comment (rjaini) [reply kubetest/azure.go]: Fixed.

File: kubetest/azure_helpers.go [hunk 5]
```diff
@@ -148,6 +150,22 @@ type AzureClient struct {
 	groupsClient      resources.GroupsClient
 }
 
+type CustomCloudProfile struct {
+	PortalURL string `json:"portalURL,omitempty"`
+}
+
+type AzureStackMetadataEndpoints struct {
```
[the remark this answers is not in the document]
Comment (rjaini) [reply kubetest/azure_helpers.go]: Fixed.

File: kubetest/azure_helpers.go [hunk 6]
```diff
@@ -148,6 +150,22 @@ type AzureClient struct {
...
+	GalleryEndpoint string                            `json:"galleryEndpoint,omitempty"`
+	GraphEndpoint   string                            `json:"graphEndpoint,omitempty"`
+	PortalEndpoint  string                            `json:"portalEndpoint,omitempty"`
+	Authentication  *AzureStackMetadataAuthentication `json:"authentication,omitempty"`
+}
+
+type AzureStackMetadataAuthentication struct {
```
[the remark this answers is not in the document]
Comment (rjaini) [reply kubetest/azure_helpers.go]: Fixed.

File: kubetest/azure_helpers.go [hunk 7]
```diff
@@ -148,6 +150,22 @@ type AzureClient struct {
 	groupsClient      resources.GroupsClient
 }
 
+type CustomCloudProfile struct {
```
[the remark this answers is not in the document]
Comment (rjaini) [reply kubetest/azure_helpers.go]: Fixed.

**Pull request merged**